### ISSUE TITLE
Putting h(3) in a fragment causes a loop in reveal.js

### ISSUE THREAD

**Issue by zareami10** [quarto-dev/quarto-cli#976]
Example:

```
---
format: revealjs
---
# Text 1

## Text 2

Text.

::: {.fragment}
### Text 3

Text.
:::

## Final Text

Text.
```

You can't see the final slide as it loops back to the starts after Text 3.

**Comment by zareami10** [comment]
Using `.column` over headings also breaks the slide in the same way.

**Comment by jjallaire** [comment]
@cderv Could you take a look at this? I can repro this behavior however if I render this same markdown with plain pandoc revealjs it doesn't repro. The generated HTML looks the same to me, perhaps a difference based on our default options?

**Comment by cderv** [comment]
I can reproduce using Pandoc by messing with `history` or `hash`, but it is not consistent and changing the option in Quarto does not help. So it may be linked to that but more complex, as a combination of options. 

However, I am not sure that the HTML produced is a correct one for RevealJS in both case. 

## TL;DR

I think the issue is with Pandoc when a fenced div is followed by a header, it will be converted to a `<section>` and not a simple `<div>`, which will mess the slide navigation within revealjs format. 

## Details

<details>
<summary>Without any fragment</summary>

````markdown
---
format: revealjs
---

# Text 1

## Text 2

Text.

### Text 3

Text

## Final Text

Text.
````

</details>

We have a `<h3>` tag created 

````html
<section>
<section id="text-1" class="title-slide slide level1">
<h1>Text 1</h1>

</section>
<section id="text-2" class="slide level2">
<h2>Text 2</h2>
<p>Text.</p>
<h3 id="text-3">Text 3</h3>
<p>Text</p>
</section>
<section id="final-text" class="slide level2">
<h2>Final Text</h2>
<p>Text.</p>
</section></section>
````

<details>
<summary>When we had the fragment fenced div we got this </summary>

````markdown
---
format: revealjs
---

# Text 1

## Text 2

Text.

::: fragment
### Text 3

Text
:::

## Final Text

Text.
````

</details>

````diff
 <section id="text-2" class="slide level2">
 <h2>Text 2</h2>
 <p>Text.</p>
-<h3 id="text-3">Text 3</h3>
+<section id="text-3" class="fragment">
+<h3>Text 3</h3>
 <p>Text</p>
 </section>
+</section>
 <section id="final-text" class="slide level2">
 <h2>Final Text</h2>
 <p>Text.</p>
````

We see that pandoc will create a `<section>` within the `<section>`, and we have no div anymore. 

I tested with `--from markdown-native_divs+raw_html` and using `<div class="fragment">` in markdown and in this case, we are keeping a `<div class="fragment">` and it works fine in both Pandoc and Quarto. 

I think the issue is with Pandoc when a fenced div is followed by a header, it will be converted to a section divs, which will mess the slide navigation within revealjs format. 

I am still looking for the conflict of option that create this behavior as it is hidden in Pandoc default, even if the HTML generated does not seem correct to me.

**Comment by cderv** [comment]
Discussing this with @tarleb, we found that pandoc is definitly handling Div + Header in some way that it will create sections with `<section>`.  This messes up revealjs. 

Suggestion is to add content between the Fenced Div dans the Header. Like a HTML comment 
````markdown
::: {.fragment}
<!-- -->
### Text 3

Text.
:::
````

or adding a id on the div like 
````markdown
::: {.fragment #q-frag-1}
### Text 3

Text.
:::
````

I am thinking that we could detect those Div and add within Quarto this "trick" to prevent the revealjs mess up. At least until this is possibly sorted out in Pandoc directly

**Comment by cderv** [comment]
I have opened https://github.com/jgm/pandoc/issues/8098 in Pandoc to discuss this further as this all comes from a Pandoc 2.8 change, that `slide-level` should affect (https://github.com/jgm/pandoc/issues/6705) but it is not working as expected

**Comment by cderv** [comment]
It seems like it will be fixed in next Pandoc version following fix for https://github.com/jgm/pandoc/issues/8098 so maybe there is no need to have a partial fix in between (like adding a Lua filter to add the comment line as in https://github.com/quarto-dev/quarto-cli/issues/976#issuecomment-1144792316)

Columns was already a special handling, so it should work already. @zareami10 interested by your example

**Comment by zareami10** [comment]
> > Using .column over headings also breaks the slide in the same way.
> 
> @zareami10 can you share the example with this ? I can't reproduce and I want to be sure the fix I have will work for this.

Unfortunately I don't have the exact setup that caused the problem (and I'm not sure whether it was limited to h2 or not), but playing around, this kind of usage completely messes up the slides:

```
---
format: revealjs
---

# Slide 1

## Slide 2

Text.

---

:::: {.columns}

::: {.column}

## Slide 3

Left side.

:::

::: {.column}

Right Side.

:::

:::


## Final Slide

Text. 
```

But to be fair I'm not sure whether putting h2 in a column would be a valid usage. But the use-case would be for example putting a vertical image on the right side of the slide spanning all the slide width.

**Comment by cderv** [comment]
> But to be fair I'm not sure whether putting h2 in a column would be a valid usage. 

Header are used to delimit slides by default. you can't use theme effectively inside divs like that. 

Right usage would be 

```
---
format: revealjs
---

# Slide 1

## Slide 2

Text.

## Slide 3

:::: {.columns}

::: {.column}

Left side.

:::

::: {.column}

Right Side.

:::

:::


## Final Slide

Text. 
````

> But the use-case would be for example putting a vertical image on the right side of the slide spanning all the slide height.

You mean having an image also going to the space of the header ? I think you would need specific CSS for that - not sure column is the right way for this. 

But you could set `slide-level: 0` to only create slides using `---` and then your code would work 

````markdown
---
format: 
  revealjs:
    slide-level: 0
---

# Slide 1

---

## Slide 2

Text.

---

:::: {.columns}

::: {.column}

## Slide 3

Left side.

:::

::: {.column}

Right Side.

:::

:::

---

## Final Slide

Text. 
````

**Comment by zareami10** [comment]
Yes that indeed works, thanks.

So my earlier issue with 'column' was probably also caused by misusing h2, but if I encounter it in other settings again I will share the markdown here.

**Comment by cderv** [comment]
I confirm the Pandoc solves this in the daily version - You can now use a header that is higher to `slide-level` inside a fenced div of class `fragment`. 

@jjallaire should we wait on a newer Pandoc or you won't include new version anyway ? I have a possible workaround for this using a Lua filter if we want to stay with last Pandoc but have this fixed. 
https://github.com/quarto-dev/quarto-cli/compare/main...cderv:reveal-div

**Comment by cderv** [comment]
@zareami10 next quarto release should have the fix for this issue. You'll be able to add headers into fragment block.

thanks a lot for the report!

**Comment by zareami10** [comment]
@cderv So I was working with slides again and noticed that a heading in `columns` causes a loop with `section-divs` set to true even when working with `slide-level: 0`.

```
---
format: revealjs
section-divs: true
slide-level: 0
---

# 1

---

:::: {.columns}
::: {.column width="50%"}
## Column 1
Text.
:::
::: {.column width="50%"}
Column 2
:::
::::

---

## 3
```

But again I'm not so sure if it's a valid use case. (section divs for example help emulating title slides when slide-level is set to 0 by allowing us use the .center class and such)

**Comment by cderv** [comment]
> with section-divs set to true

I really don't think that `--sections-divs` should be used with `revealjs` format. `<section>` is used by Reveal to delimit some slides and using section divs will add more `<section>` that won't have that meaning. So it will indeed mess up the slides IMO. I would not use that at all. 

> section divs for example help emulating title slides when slide-level is set to 0 by allowing us use the .center class and such

Don't use `slide-level = 0` in this case then. It is still indeed not possible to add attributes on such element as `---` (until this is allowed https://github.com/jgm/pandoc/issues/684)

Note that you can create empty title slide if you need them 
````
---
format: revealjs
---

# 1

## {.center}

:::: {.columns}
::: {.column width="50%"}
### Column 1
Text.
:::
::: {.column width="50%"}
Column 2
:::
::::

## 3
````
